util/reward-calculator/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+
+#[derive(Debug, PartialEq, Clone, Eq, Fail)]
+pub enum Error {
+    #[fail(display = "Can't resolve finalize target: {}", _0)]
+    Target(BlockNumber),
+    #[fail(display = "Can't parse Script from target witness: {:x}", _0)]
+    Script(H256),
+}
+
+pub struct RewardCalculator<'a, P> {
+    pub provider: &'a P,
+}
+
+impl<'a, P: ChainProvider> RewardCalculator<'a, P> {
+    pub fn new(provider: &'a P) -> Self {
+        RewardCalculator { provider }
+    }
+
+    pub fn block_reward(&self, parent: &Header) -> Result<(Script, Capacity), FailureError> {
+        let consensus = self.provider.consensus();
+        let store = self.provider.store();
+
+        let block_number = parent.number() + 1;
+        let target_number = consensus
+            .finalize_target(block_number)
+            .ok_or_else(|| Error::Target(block_number))?;
+
+        let target = self
+            .provider
+            .get_ancestor(parent.hash(), target_number)
+            .ok_or_else(|| Error::Target(block_number))?;
+
+        let target_lock = Script::from_witness(
+            &store
+                .get_cellbase(target.hash())
+                .expect("target cellbase exist")
+                .witnesses()[0],
+        )
+        .ok_or_else(|| Error::Script(target.hash().to_owned()))?;
+
+        let txs_fees = self.txs_fees(&target)?;
+        let proposal_reward = self.proposal_reward(parent, &target)?;
+        let base_block_reward = self.base_block_reward(&target)?;
+
+        let reward = txs_fees
+            .safe_add(proposal_reward)?
+            .safe_add(base_block_reward)?;
+        Ok((target_lock, reward))
+    }
+
+    pub fn txs_fees(&self, target: &Header) -> Result<Capacity, FailureError> {
+        let target_ext = self
+            .provider
+            .store()
+            .get_block_ext(target.hash())
+            .expect("block body stored");
+
+        target_ext
+            .txs_fees
+            .iter()
+            .try_fold(Capacity::zero(), |acc, tx_fee| {
+                tx_fee.safe_mul_ratio(4, 10).and_then(|proposer| {
+                    tx_fee
+                        .safe_sub(proposer)
+                        .and_then(|miner| acc.safe_add(miner))
+                })
+            })
+            .map_err(Into::into)
+    }
+
+    pub fn proposal_reward(
+        &self,
+        parent: &Header,
+        target: &Header,
+    ) -> Result<Capacity, FailureError> {
+        let mut target_proposals = self.get_proposal_ids_by_hash(target.hash());
+
+        let proposal_window = self.provider.consensus().tx_proposal_window();
+        let block_number = parent.number() + 1;
+        let store = self.provider.store();
+
+        let mut reward = Capacity::zero();
+        let commit_start = cmp::max(block_number.saturating_sub(proposal_window.length()), 2);
+        let proposal_start = cmp::max(commit_start.saturating_sub(proposal_window.start()), 1);
+
+        let mut proposal_table = BTreeMap::new();
```

Comment:
I propose a simpler algorithm below.

We only need to keep a single `HashSet`, which tracks the proposal ids that have already been proposed in previous blocks. This can avoid repeat 8 `flat_map`s.

We name the `HashSet` as `proposed`, which stores txs already proposed before the target block.

The loop starts from parent block, backward to `commit_start`. For each block `bn`:

- Find proposals of `bn - 10`, add the proposals to `proposed` unless `bn - 10 == target`.
- For each tx in block bn, if tx is not in `proposed` but in target's proposal, sum the proposal reward.

Reply:
Alternatively, the proposed ids can be removed from target_proposals, no need to maintain another hash set.